facilitator/Dockerfile

```diff
@@ -1,16 +1,39 @@
 FROM rust:1.46-alpine as builder
 
 RUN apk add libc-dev && apk update
+# Attempt to install a nonexistent package. This triggers
+# updating the crates.io index separately from building the
+# dependencies, so if dependencies change we don't have to
+# re-download the whole index.
+RUN cargo install _update_crates_io_failure_is_expected_ ; true
```

Comment:
I'm surprised that `cargo` gets as far as installing all the dependencies before failing because the target doesn't exist. Is that documented behavior of `cargo`?

Reply:
Crates has an "index" - a big git repository containing a JSON file with metadata for every crate in the ecosystem. It downloads the whole thing on first run, and then updates it periodically. This command triggers the first fetch of that index (but not any dependency installs). Otherwise the first fetch would run with the first `cargo build`. It's perhaps a bit of an over-optimization -- it makes full builds when you rebuild the dependencies faster -- but I think it's fine.

The behavior that `cargo install` fetches the index is not documented but is a necessary implication of the design. Cargo can't tell that a package is nonexistent until it has a copy of the index.